### PR TITLE
React Select v2

### PR DESCRIPTION
**Pull request by JedWatson**
**Updated:** See the preview here: http://bit.ly/react-select-v2

v2 issue for discussion: #2208

View the v2 source: https://github.com/JedWatson/react-select/tree/v2

Alpha releases of react-select@2.0.0 are available on npm for preview, and I have opened this PR to help increase visibility of the current state, ci status and netlify preview.

v2 should still be considered experimental, but it's at a point where I would absolutely welcome feedback on the architecture and approach. More examples and documentation coming soon.

PRs and issues specific to the new version should be tagged `v2`

### CODE REVIEW

**Comment by JedWatson** [comment]
I just published `react-select@2.0.0-alpha.2` to npm, this is getting pretty close to a viable implementation. The examples have been updated (see preview link above) and I'd love some feedback on the new architecture.

v2 significantly improves on v1 in several areas:

- Uses css-in-js with a complete styling API
- Allows you to completely replace any of the built-in rendering components
- Is simpler and more extensible (many props will become redundant)
- More performant

Major work to do before v2 hits feature parity with v1 includes:

- [ ] Form integration (hidden inputs)
- [ ] Async options with Promises (currently supports callbacks)
- [ ] Creatable (tags) mode
- [ ] More robust filtering functionality
- [ ] Virtualisation support
- [ ] Ability to customise built-in strings

It also needs complete documentation and an upgrade guide with changes from v1.x, as well as a lot of tests (and user testing). In the meantime, if you're interested read the [example source](https://github.com/JedWatson/react-select/tree/v2/examples/pages) to see how the new API works.

**Comment by JedWatson** [comment]
Quick update: I published `v2.0.0-alpha.3` today and have written up a more comprehensive to-do list [here](https://github.com/JedWatson/react-select/blob/v2/.TODO.md)

Getting closer!

**Comment by Zertz** [comment]
Amazing work! We're starting a new project using v2, looking great so far 💯 

Only thing so far is that I can't seem to get the `loadOptions` to work with a Promise but the callback method works just fine.

**Comment by JedWatson** [comment]
@Zertz great to hear! I've just implemented Promise support in Async, it'll be available in the upcoming `alpha.4` release

@albertogasparin thanks for the review 👌 some good points in there. I'm currently working on a pretty big cleanup (see #2351) and will work through your feedback as soon as I land that.

File: src/components/Input.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+type Props = PropsWithStyles & PropsWithInnerRef & { isHidden: boolean };
+
+export const css = () => marginHorizontal(spacing.baseUnit / 2);
+const inputStyle = isHidden => ({
+  background: 0,
+  border: 0,
+  fontSize: 'inherit',
+  opacity: isHidden ? 0 : 1,
+  outline: 0,
+  padding: 0,
```
[the remark this answers is not in the document]
Comment (JedWatson) [reply src/components/Input.js]: @jossmac can you take a look at this?

File: src/Select.js [hunk 2]
```diff
@@ -1,1316 +1,1052 @@
...
+        isDisabled,
+        isSelected,
+        key: `option-${value}`,
+        label,
+        type: 'option',
+        value,
+      };
+    };
+
+    options.forEach((item, itemIndex) => {
```
[the remark this answers is not in the document]
Comment (JedWatson) [reply src/Select.js]: @albertogasparin would you be interested in creating a PR for that?

File: src/Async.js [hunk 3]
```diff
@@ -1,210 +1,147 @@
...
+      this.setState(
+        {
+          inputValue,
+          isLoading: true,
+          passEmptyOptions: !this.state.loadedInputValue,
+        },
+        () => {
+          this.loadOptions(inputValue, options => {
+            if (!this.mounted) return;
+            if (options) {
```
[the remark this answers is not in the document]
Comment (JedWatson) [reply src/Async.js]: Yeah that seems like a good thing to do

File: src/Async.js [hunk 4]
```diff
@@ -1,210 +1,147 @@
...
+      inputValue: '',
+      isLoading: props.defaultOptions === true ? true : false,
+      loadedOptions: [],
+      passEmptyOptions: false,
+    };
+  }
+  componentDidMount() {
+    this.mounted = true;
+    const { defaultOptions } = this.props;
+    if (defaultOptions === true) {
```
[the remark this answers is not in the document]
Comment (JedWatson) [reply src/Async.js]: I think you're right, it _does_ seem a bit less predictable this way. Happy to add another prop.

File: .TODO.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+* [ ] * `searchable` / `searchPromptText`
+* [ ] * `tabIndex`
+
+---
+
+# Maybe:
+
+* [ ] Prevent values from being popped, was `option.clearableValue === false`
+* [ ] Scroll behaviour: should we detect parent? how do we handle the footer?
+* [ ] Scroll behaviour: can we overscroll up to show the group heading?
```
[the remark this answers is not in the document]
Comment (JedWatson) [reply .TODO.md]: Yeah, we should also do some work around handling more complex requirements for rendering the menu (e.g into a Portal)

File: src/components/Input.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (jossmac) [reply src/components/Input.js]: @JedWatson sure thing, I'll check it out tomorrow.

[the remark this answers is not in the document]
Comment (jossmac) [reply src/components/Input.js]: @albertogasparin @JedWatson resolved in 07bba99 of #2354

File: src/Select.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertogasparin) [reply src/Select.js]: Sure. Here is my take #2359

**Comment by Zertz** [comment]
Fantastic work @JedWatson ! We're releasing `alpha.6` in production soon :)

**Comment by shulcsm** [comment]
Could you release an alpha, it's been a while.

**Comment by barbu110** [comment]
What about TypeScript typings?

BTW This is an amazing project!

**Comment by JedWatson** [comment]
Heads up everyone who's watching this thread - `alpha.9` was just released, and we'll be moving into beta shortly (at which point this will be merged into `master`)

@barbu110 

> What about TypeScript typings

I'll consider that after the final release of 2.0.0, I know they're helpful but I'm reluctant to introduce anything that needs to be kept in sync after the experience of maintaining both less and scss in v1.x, where only less was actually used by the examples and build. I'm not sure what's worse between no types, and out of sync types.

**Comment by bcalik** [comment]
Thanks for the great effort @JedWatson !! v2 looks really promising. I can't wait to use it on my next project.

**Comment by falconmick** [comment]
Universal/SSR is broken due to https://github.com/JedWatson/react-select/blob/v2/src/utils.js line 206

I am pulling down locally and working on a patch, will submit PR as soon as I've got SSR working

**Comment by JedWatson** [comment]
Great news - React Select 2.0.0-beta.1 is out 🎉 

At this point, I think the rewrite is basically complete and don't expect much to change before general release of 2.0. So, I need help making sure it's ready 😃 

I'd like to keep it in beta for a couple of weeks, so there's time to get feedback and make sure the upgrade guide / docs are ready before we merge it into `master`, but all the features I'm planning for 2.0.0 are ready and it's well tested.

Thanks for picking up SSR @falconmick, if anyone else finds issues during beta please open an issue (or even better, PR) with the prefix [v2] in the title.

**Comment by pachuka** [comment]
Is it possible to still use css classes instead of css-in-js? Also the @types/react-select will need to be updated for those using this with Typescript. Other than that this looks awesome!

**Comment by falconmick** [comment]
@JedWatson I'm having a super hard time building the project so I can test it locally. I'm running:

yarn build
yarn link

go into my project

yarn link react-select

First time I did this, my build uses the "main" defined in package.json which resolves to the lib folder and I get the following error:

```
Error: Cannot find module "glam"
```

So I go and I move that main to instead point at build/react-select.js, but then I get this error: 

```
ReferenceError: createFilter is not defined
```

Finally I try build/react-select.es.js

```
SyntaxError: Unexpected token import
```

How exactly can I build and test this locally via yarn link. This is really painful to work with.

File: docs/pages/styles/index.js [hunk 6]
```diff
@@ -104,6 +104,32 @@ export default function Styles() {
           <StyledMulti />
         </ExampleWrapper>
       )}
+
+    ## Using classNames
+
+    If you provide the \`className\` prop to react-select, all inner elements will
```
Comment (eemeli) [inline docs/pages/styles/index.js]: Shouldn't this be `classNamePrefix`?

**Comment by falconmick** [comment]
@JedWatson where can I see a list of remaining tasks that are small enough for new-ish contributers

**Comment by dbertella** [comment]
Dunno if it's the right space to discuss, but I'm wondering why did you go for a new prop `isDisabled` instead of sticking to standard `disabled`. This will be a little pain for the migration

**Comment by jossmac** [comment]
@dbertella from the initial comment:

> v2 issue for discussion: #2208

* * *

The `value` property _has_ remained the same; it's likely that you were using `simpleValue`.

**Comment by JedWatson** [comment]
OMGF 🎉

**Pull request merged**